draft-ietf-jsonpath-base.md

```diff
@@ -413,51 +436,77 @@ the grammar in this document, its UTF-8 form SHOULD first be decoded into
-
-For any implementation, if a syntactically invalid JSONPath is provided,
-the implementation MUST return an error.
-
-If a syntactially invalid JSON value is provided, any implementation SHOULD
-return an error.
+A string to be used as a JSONPath query needs to be *well-formed* and
+*valid*.
+A string is a well-formed JSONPath query if it conforms to the syntax
+of JSONPath.
+A well-formed JSONPath query is valid if it also fulfills all semantic
+requirements posed by this document.
+
+The well-formedness and the validity of JSONPath queries are independent of
+the JSON value the query is applied to; no further errors can be
+raised during application of the query to a JSON value.
+
+(Obviously, an implementation can still fail when executing a JSONPath
+query, e.g., because of resource depletion, but this is not modeled in
+the present specification.)
+
+> Discussion: are we ready for the error model defined here, i.e. one
+> that does not have runtime errors, but only compile time errors?
+
+## Processing Model
+
+In this specification, the semantics of a JSONPath query are defined
+in terms of a *processing model*.  That model is not prescriptive of
+the internal workings of an implementation:  Implementations may wish
+(or need) to design a different process that yields results that
+conform to the model.
+
+In the processing model,
+a valid JSONPath query is executed against a JSON value, the *Argument*, and
+produces a list of zero or more nodes of the JSON value which are
+selected by the JSONPath.
+(Note that the term JSON value also implies that this value complies
+to the definitions in {{-json}}, i.e., is indeed a JSON value.)
+
+The JSONPath query is a sequence of zero or more *selectors*, each of
+which is applied to the result of the previous selector and leaves
+input to the next selector.
+These results and inputs take the form of a *nodelist*, i.e., a
+sequence of zero or more nodes.
+
+The nodelist going into the first selector contains a single node,
+holding the Argument.
+The nodelist resulting from the last selector is presented as the
+result of the query; depending on the specific API, it might be
+presented as an array of the JSON values at the nodes, an array of
+Output Paths referencing the nodes, or both — or some other
+representation as desired by the implementation.
+Note that the API must be capable of presenting an empty nodelist as
+the result of the query.
+
+A selector performs its function on each of the nodes in its input
+nodelist, during such a function execution, such a node is referred to
+as "current node".  Each of these function executions produces a
+result nodelist, which are then combined by algorithm *combine1* into
+the result of the selector.
+
+> Discussion: We haven't decided yet whether there is only one
+> *combine1* or a choice, with for instance simple concatenation and
+> concatenation with duplicate removal, based on a definition of
+> duplicate, are candidates.
```

Comment:
I think a single *combine1* is sufficient and much simpler than allowing selector-specific behaviour.